templates/puppet/integration_spec.rb

```diff
@@ -0,0 +1,194 @@
+environment and loader.  Since these functions don't call out to Facter
+the way that `puppet apply` does, we need to substitute in our own project
+name and credential path (which all our examples use).  Hence the
+get_example() function - it does more than just read the file.
+
+There's also the issue of the begin/rescue pattern you'll see here.  It's
+ugly!  It's also necessary.  compile_to_ral() calls out to our mocked up
+loader, but fails to successfully compile the manifest due to some
+initialization which is mandatory, but isn't possible to complete from
+outside puppet.  HOWEVER!  We're in luck!  In calling some cleanup code,
+in reporting on the failure, compile_to_ral actually *performs* the
+initialization we need.
+
+So!  We *can* successfully apply a manifest from
+under webmock, as long as we:
+- try twice, eating the first failure.
+- don't need to use any functions that are loaded from modules
+- ensure all our dependent modules have their 'lib' dir in $LOAD_PATH
+- do all our own variable substitution
+- rewrite any necessary functions in puppet and stick them in
+    the manifest.
+- import all the magic from puppetlabs_spec_helper/module_spec_helper,
+    which I have done without fully understanding it.  The comments in
+    the code suggest that the authors also do not fully understand it.
+
+Hopefully that explains the complexity here in a way that will continue
+to be useful as the tests develop further.
+
+Let me also explain the order in which these commands are run.  First,
+there is a pre-destroy - in the event that an older test failed to clean
+up after itself, the next run should take care of it.  Second, we create
+the resources in the manifest.  Third, we run the create again, mocking
+out any requests during 'flush' and making them crash the test - this is
+to check idempotency of the manifests.  Fourth, we delete the resources
+in the manifest, and fifth, we check the idempotency of that delete by
+confirming that no resources make web requests during 'flush'.
+
+VCR recordings are called 'cassettes', and they are stored in
+'spec/cassettes'.  If present, they are used - if absent, created.
+#-%>
+<%= compile 'templates/license.erb' -%>
+
+<%= lines(autogen_notice :puppet) -%>
+require 'spec_helper'
+require 'vcr'
+
+VCR.configure do |c|
+  c.cassette_library_dir = 'spec/cassettes'
+  c.hook_into :webmock
+  c.configure_rspec_metadata!
+end
+
+describe '<%= Google::StringUtils.underscore(obj.name) -%>.create', vcr: true do
+  mods = [File.expand_path('.'), File.expand_path('./spec/fixtures/modules/')]
+  it 'creates and destroys non-existent <%= Google::StringUtils.underscore(obj.name) -%>' do
+    puts 'pre-destroying <%= Google::StringUtils.underscore(obj.name) -%>'
+    VCR.use_cassette('pre_destroy_<%= Google::StringUtils.underscore(obj.name) -%>') do
+      example = get_example('delete_<%= Google::StringUtils.underscore(obj.name) -%>')
+      test = Puppet::Node::Environment.create(:test, mods)
```

Comment:
Is there any way we can hide lines 110-118 (ish) in a function somewhere?

It would make the tests more readable at a higher-level

Reply:
Yep, done.